logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java

```diff
@@ -0,0 +1,196 @@
+      isEnqueued =
+          queue.offer(new SequencedBatch(currentPosition, sourcePosition, List.of(appendEntry)));
+      if (isEnqueued) {
+        position = currentPosition + 1;
+      }
+    } finally {
+      lock.unlock();
+    }
+
+    if (consumer != null) {
+      consumer.signal();
+    }
+    metrics.observeBatchSize(1);
+    metrics.setQueueSize(queue.size());
+    if (isEnqueued) {
+      return currentPosition;
+    } else {
+      LOG.trace("Rejecting write of {}, sequencer queue is full", appendEntry);
+      return -1;
+    }
+  }
+
+  /**
+   * @param appendEntries a set of entries to append; these will be appended in the order in which
+   *     the collection is iterated.
+   * @param sourcePosition a back-pointer to the record whose processing created these entries
+   * @return -1 if write was rejected, 0 if batch was empty, the highest position of the batch if
+   *     write was successful.
+   */
+  @Override
+  public long tryWrite(
+      final Iterable<? extends LogAppendEntry> appendEntries, final long sourcePosition) {
+    if (isClosed) {
+      LOG.warn("Rejecting write of {}, sequencer is closed", appendEntries);
+      return -1;
+    }
+
+    final var entries = new ArrayList<LogAppendEntry>(16);
+    for (final var entry : appendEntries) {
+      entries.add(entry);
+    }
+    final var batchSize = entries.size();
+    if (batchSize == 0) {
+      return 0;
+    }
+
+    final long currentPosition;
+    final boolean isEnqueued;
+    lock.lock();
+    try {
+      currentPosition = position;
+      isEnqueued = queue.offer(new SequencedBatch(currentPosition, sourcePosition, entries));
+      if (isEnqueued) {
+        position = currentPosition + batchSize;
+      }
+    } finally {
+      lock.unlock();
+    }
+
+    if (consumer != null) {
+      consumer.signal();
+    }
+    metrics.setQueueSize(queue.size());
+    if (isEnqueued) {
+      metrics.observeBatchSize(batchSize);
+      return currentPosition + batchSize - 1;
+    } else {
+      LOG.trace("Rejecting write of {}, sequencer queue is full", entries);
+      return -1;
+    }
+  }
+
+  /**
+   * Tries to read a {@link SequencedBatch} from the queue.
+   *
+   * @return A {@link SequencedBatch} or null if none is available.
+   */
+  public SequencedBatch tryRead() {
+    return queue.poll();
+  }
+
+  public SequencedBatch peek() {
+    return queue.peek();
+  }
+
+  /**
+   * Closes the sequencer. After closing, writes are rejected but reads are still allowed to drain
+   * the queue. Closing the sequencer is not atomic so some writes may occur shortly after closing.
+   */
+  @Override
+  public void close() {
+    LOG.info("Closing sequencer for writing");
+    isClosed = true;
+  }
+
+  /**
+   * @return true if the sequencer is closed for writing.
+   */
+  public boolean isClosed() {
+    return isClosed;
+  }
+
+  public void registerConsumer(final ActorCondition consumer) {
+    this.consumer = consumer;
+    consumer.signal();
```

Comment:
:x: Shouldn't we signal only if the queue is not empty? I guess it doesn't hurt anyway :shrug: The other option is to have the consumer register itself and immediately poll for work :shrug:

Reply:
My reasoning was "doesn't hurt" but of course we could flip it and let the appender start polling. Does it matter?

Reply:
I think I would prefer the consumer to poll immediately, if only to avoid extra work of going through the actor scheduler via signaling. Once registered, the consumer WILL be signaled if new work comes in, so if it polls immediately and nothing's there, it won't lose anything (I think? :sweat_smile:)

But super minor, so let's make this a soft suggestion :smile: This approach is definitely more defensive, so I'm fine with either :+1:

Reply:
:+1: I implemented your suggestion.